plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtRestoreVMSnapshotCommandWrapper.java

```diff
@@ -0,0 +1,121 @@
+import com.cloud.resource.CommandWrapper;
+import com.cloud.resource.ResourceWrapper;
+import com.cloud.utils.script.Script;
+import com.cloud.vm.VirtualMachine;
+
+@ResourceWrapper(handles =  RestoreVMSnapshotCommand.class)
+public final class LibvirtRestoreVMSnapshotCommandWrapper extends CommandWrapper<RestoreVMSnapshotCommand, Answer, LibvirtComputingResource> {
+
+    private static final Logger s_logger = Logger.getLogger(LibvirtRestoreVMSnapshotCommandWrapper.class);
+
+    @Override
+    public Answer execute(final RestoreVMSnapshotCommand cmd, final LibvirtComputingResource libvirtComputingResource) {
+        String vmName = cmd.getVmName();
+        List<VolumeObjectTO> listVolumeTo = cmd.getVolumeTOs();
+        VirtualMachine.PowerState vmState = VirtualMachine.PowerState.PowerOn;
+
+        Domain dm = null;
+        try {
+            final LibvirtUtilitiesHelper libvirtUtilitiesHelper = libvirtComputingResource.getLibvirtUtilitiesHelper();
+            Connect conn = libvirtUtilitiesHelper.getConnection();
+            dm = libvirtComputingResource.getDomain(conn, vmName);
+
+            if (dm == null) {
+                return new RestoreVMSnapshotAnswer(cmd, false,
+                        "Restore VM Snapshot Failed due to can not find vm: " + vmName);
+            }
+            String xmlDesc = dm.getXMLDesc(0);
+
+            List<VMSnapshotTO> snapshots = cmd.getSnapshots();
+            Map<Long, VMSnapshotTO> snapshotAndParents = cmd.getSnapshotAndParents();
+            for (VMSnapshotTO snapshot: snapshots) {
+                VMSnapshotTO parent = snapshotAndParents.get(snapshot.getId());
+                String parentName = (parent == null)? "": ("  <parent><name>" + parent.getSnapshotName() + "</name></parent>\n");
+                String vmSnapshotXML = "<domainsnapshot>\n"
+                        + "  <name>" + snapshot.getSnapshotName() + "</name>\n"
+                        + "  <state>running</state>\n"
+                        + parentName
+                        + "  <creationTime>" + (int) Math.rint(snapshot.getCreateTime()/1000) + "</creationTime>\n"
+                        + xmlDesc
+                        + "</domainsnapshot>";
+                s_logger.debug("Restoring vm snapshot " + snapshot.getSnapshotName() + " on " + vmName + " with XML:\n " + vmSnapshotXML);
+                File tmpCfgFile = null;
+                try {
+                    tmpCfgFile = File.createTempFile(vmName + "-", "cfg");
+                    final PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter(tmpCfgFile)));
+                    out.println(vmSnapshotXML);
+                    out.close();
+                    String cfgFilePath = tmpCfgFile.getAbsolutePath();
+                    String cmdvirsh = "virsh snapshot-create --redefine " + vmName + (snapshot.getCurrent()? " --current ":" ") + cfgFilePath;
+                    int cmdout = Script.runSimpleBashScriptForExitValue(cmdvirsh);
```

Comment:
Although I understand that the libvirt-java bindings don't have all the required support, we should not introduce any new cases where we execute "virsh".

The proper way to do this would be to implement these features in libvirt-java and depend on a new release of those bindings.